#! /bin/sh -e
## DP: Adds support for drac
## DP: Author: Timo Sirainen

# Based on patch from http://dovecot.org/patches/drac.c

. $(dirname $0)/DPATCH

exit 0
@DPATCH@
diff -urN dovecot-1.1.1/configure.in dovecot-1.1.1-drac/configure.in
--- dovecot-1.1.1/configure.in	2008-06-22 05:02:27.000000000 -0600
+++ dovecot-1.1.1-drac/configure.in	2008-06-23 21:12:49.000000000 -0600
@@ -2243,6 +2243,7 @@
 src/plugins/mbox-snarf/Makefile
 src/plugins/quota/Makefile
 src/plugins/imap-quota/Makefile
+src/plugins/drac/Makefile
 src/plugins/trash/Makefile
 src/plugins/zlib/Makefile
 stamp.h
diff -urN dovecot-1.1.1/src/plugins/drac/drac.c dovecot-1.1.1-drac/src/plugins/drac/drac.c
--- dovecot-1.1.1/src/plugins/drac/drac.c	1969-12-31 17:00:00.000000000 -0700
+++ dovecot-1.1.1-drac/src/plugins/drac/drac.c	2008-06-22 20:39:12.000000000 -0600
@@ -0,0 +1,60 @@
+/* Copyright (C) 2003 Timo Sirainen, LGPLv2.1 */
+
+/* Installation:
+
+   export dovecot=~/src/dovecot
+
+   gcc -Wall -W -shared -fPIC -DHAVE_CONFIG_H -I$dovecot -I$dovecot/src/lib \
+   drac.c -o drac.so -ldrac
+
+   cp drac.so /usr/local/lib/dovecot/imap/
+   cp drac.so /usr/local/lib/dovecot/pop3/
+*/
+
+#include "lib.h"
+#include "ioloop.h"
+#include "network.h"
+
+#include <stdlib.h>
+
+#define DRAC_TIMEOUT_SECS 60
+#define DRAC_HOST "localhost"
+
+int dracauth(char *host, unsigned long ip, char **errmsg);
+
+static struct timeout *to_drac = NULL;
+static unsigned long in_ip;
+
+static void drac_timeout(void *context ATTR_UNUSED)
+{
+	char *err;
+
+	if (dracauth(DRAC_HOST, in_ip, &err) != 0)
+		i_error("dracauth() failed: %s", err);
+}
+
+void drac_init(void)
+{
+	const char *ip_str;
+	struct ip_addr ip;
+
+	ip_str = getenv("IP");
+	if (ip_str == NULL)
+		i_error("DRAC: IP environment not given");
+	else if (net_addr2ip(ip_str, &ip) < 0)
+		i_error("DRAC: net_ip2addr(%s) failed: %m", ip_str);
+	else if (!IPADDR_IS_V4(&ip))
+		i_error("DRAC: Only IPv4 addresses are supported (%s)", ip_str);
+	else {
+		in_ip = ((struct in_addr *) &ip.u.ip4)->s_addr;
+		drac_timeout(NULL);
+		to_drac = timeout_add(1000*DRAC_TIMEOUT_SECS,
+				      drac_timeout, NULL);
+	}
+}
+
+void drac_deinit(void)
+{
+	if (to_drac != NULL)
+		timeout_remove(&to_drac);
+}
diff -urN dovecot-1.1.1/src/plugins/drac/Makefile.am dovecot-1.1.1-drac/src/plugins/drac/Makefile.am
--- dovecot-1.1.1/src/plugins/drac/Makefile.am	1969-12-31 17:00:00.000000000 -0700
+++ dovecot-1.1.1-drac/src/plugins/drac/Makefile.am	2008-06-23 22:39:24.000000000 -0600
@@ -0,0 +1,20 @@
+AM_CPPFLAGS = \
+	-I$(top_srcdir)/src/lib
+
+lib11_drac_plugin_la_LDFLAGS = -module -avoid-version
+
+module_LTLIBRARIES = \
+	lib11_drac_plugin.la
+
+lib11_drac_plugin_la_SOURCES = \
+	drac.c
+
+#noinst_HEADERS = \
+#	drac.h
+
+install-exec-local:
+	for d in imap pop3; do \
+	  $(mkdir_p) $(DESTDIR)$(moduledir)/$$d; \
+	  rm -f $(DESTDIR)$(moduledir)/$$d/lib11_drac_plugin.so; \
+	  $(LN_S) ../lib11_drac_plugin.so $(DESTDIR)$(moduledir)/$$d; \
+	done
diff -urN dovecot-1.1.1/src/plugins/Makefile.am dovecot-1.1.1-drac/src/plugins/Makefile.am
--- dovecot-1.1.1/src/plugins/Makefile.am	2008-06-12 00:45:10.000000000 -0600
+++ dovecot-1.1.1-drac/src/plugins/Makefile.am	2008-06-23 23:08:32.000000000 -0600
@@ -7,6 +7,6 @@
 endif
 
 SUBDIRS = \
-	acl convert expire fts fts-squat lazy-expunge mail-log mbox-snarf \
+	acl convert expire fts fts-squat lazy-expunge mail-log mbox-snarf drac \
 	quota imap-quota trash \
 	$(ZLIB) $(FTS_LUCENE)
